vector/src/main/java/im/vector/riotx/features/terms/ReviewTermsViewModel.kt

```diff
@@ -0,0 +1,157 @@
+        override fun create(viewModelContext: ViewModelContext, state: ReviewTermsViewState): ReviewTermsViewModel? {
+            val activity: ReviewTermsActivity = (viewModelContext as ActivityViewModelContext).activity()
+            return activity.viewModelFactory.create(state)
+        }
+    }
+
+    lateinit var termsArgs: ServiceTermsArgs
+
+    override fun handle(action: ReviewTermsAction) {
+        when (action) {
+            is ReviewTermsAction.LoadTerms          -> loadTerms(action)
+            is ReviewTermsAction.MarkTermAsAccepted -> markTermAsAccepted(action)
+            ReviewTermsAction.Accept                -> acceptTerms()
+        }.exhaustive
+    }
+
+    private fun markTermAsAccepted(action: ReviewTermsAction.MarkTermAsAccepted) = withState { state ->
+        val newList = state.termsList.invoke()?.map {
+            if (it.url == action.url) {
+                it.copy(accepted = action.accepted)
+            } else {
+                it
+            }
+        }
+
+        if (newList != null) {
+            setState {
+                state.copy(
+                        termsList = Success(newList)
+                )
+            }
+        }
+    }
+
+    private fun acceptTerms() = withState { state ->
+        val acceptedTerms = state.termsList.invoke() ?: return@withState
+
+        if (acceptedTerms.any { it.accepted.not() }) {
+            // Should not happen
+            _viewEvents.post(ReviewTermsViewEvents.Failure(IllegalStateException("Please accept all terms"), false))
+            return@withState
+        }
+
+        setState {
+            copy(
+                    acceptingTerms = Loading()
+            )
+        }
+
+        val agreedUrls = acceptedTerms.map { it.url }
+
+        session.agreeToTerms(
+                termsArgs.type,
+                termsArgs.baseURL,
+                agreedUrls,
+                termsArgs.token,
+                object : MatrixCallback<Unit> {
+                    override fun onSuccess(data: Unit) {
+                        _viewEvents.post(ReviewTermsViewEvents.Success)
+                    }
+
+                    override fun onFailure(failure: Throwable) {
+                        Timber.e(failure, "Failed to agree to terms")
+                        setState {
+                            copy(
+                                    acceptingTerms = Uninitialized
+                            )
+                        }
+                        _viewEvents.post(ReviewTermsViewEvents.Failure(failure, false))
+                    }
+                }
+        )
+    }
+
+    private fun loadTerms(action: ReviewTermsAction.LoadTerms) = withState { state ->
+        if (state.termsList is Loading || state.termsList is Success) {
+            return@withState
+        }
+
+        setState {
+            copy(termsList = Loading())
+        }
+
+        session.getTerms(termsArgs.type, termsArgs.baseURL, object : MatrixCallback<GetTermsResponse> {
```

Comment:
viewModelScope.launch?

Reply:
Yes, but we should perform this change globally in the app, right?